openmdao/core/driver.py

```diff
@@ -106,19 +117,115 @@ def _setup_driver(self, problem):
         self._problem = problem
         model = problem.model
 
+        self._objs = objs = OrderedDict()
+        self._cons = cons = OrderedDict()
+        self._responses = model.get_responses(recurse=True)
+        for name, data in iteritems(self._responses):
+            if data['type'] == 'con':
+                cons[name] = data
+            else:
+                objs[name] = data
+
         # Gather up the information for design vars.
         self._designvars = model.get_design_vars(recurse=True)
 
-        self._responses = model.get_responses(recurse=True)
-        self._objs = model.get_objectives(recurse=True)
-        self._cons = model.get_constraints(recurse=True)
+        con_set = set()
+        obj_set = set()
+        dv_set = set()
+
+        self._remote_dvs = dv_dict = {}
+        self._remote_cons = con_dict = {}
+        self._remote_objs = obj_dict = {}
+
+        # Now determine if later we'll need to allgather cons, objs, or desvars.
+        if model.comm.size > 1 and model._subsystems_allprocs:
+            local_out_vars = set(model._outputs._views)
+            remote_dvs = set(self._designvars) - local_out_vars
+            remote_cons = set(self._cons) - local_out_vars
+            remote_objs = set(self._objs) - local_out_vars
+            all_remote_vois = model.comm.allgather((remote_dvs, remote_cons, remote_objs))
+            for rem_dvs, rem_cons, rem_objs in all_remote_vois:
+                con_set.update(rem_cons)
+                obj_set.update(rem_objs)
+                dv_set.update(rem_dvs)
+
+            # If we have remote VOIs, pick an owning rank for each and use that
+            # to bcast to others later
+            owning_ranks = model._owning_rank['output']
+            sizes = model._var_sizes['nonlinear']['output']
+            for i, vname in enumerate(model._var_allprocs_abs_names['output']):
+                owner = owning_ranks[vname]
+                if vname in dv_set:
+                    dv_dict[vname] = (owner, sizes[owner, i])
+                if vname in con_set:
+                    con_dict[vname] = (owner, sizes[owner, i])
+                if vname in obj_set:
+                    obj_dict[vname] = (owner, sizes[owner, i])
+
+        self._remote_responses = self._remote_cons.copy()
```

Comment:
Missed `self._remote_responses` in the attributes section of the class docstring

Reply:
fixed